pylsp_mypy/plugin.py

```diff
@@ -90,25 +90,24 @@ def parse_line(line: str, document: Optional[Document] = None) -> Optional[Dict[
 
         lineno = int(linenoStr or 1) - 1  # 0-based line number
         offset = int(offsetStr or 1) - 1  # 0-based offset
+        end_lineno = (int(endlinenoStr) - 1) if endlinenoStr else None
+        end_offset = (int(endoffsetStr)) if endoffsetStr else None
         errno = 2
         if severity == "error":
             errno = 1
         diag: Dict[str, Any] = {
             "source": "mypy",
             "range": {
                 "start": {"line": lineno, "character": offset},
-                # There may be a better solution, but mypy does not provide end
-                "end": {"line": lineno, "character": offset + 1},
+                "end": {"line": end_lineno or lineno, "character": end_offset},
             },
             "message": msg,
             "severity": errno,
         }
-        if document:
-            # although mypy does not provide the end of the affected range, we
-            # can make a good guess by highlighting the word that Mypy flagged
-            word = document.word_at_position(diag["range"]["start"])
-            if word:
-                diag["range"]["end"]["character"] = diag["range"]["start"]["character"] + len(word)
+        if diag["range"]["end"]["character"] is None:
+            if document:
+                word = document.word_at_position(diag["range"]["start"])
+                diag["range"]["end"]["character"] = offset + len(word) if word else offset + 1
```

Comment:
Maybe move that code to after where `end_offset` is defined instead of adjusting diagnostic after creating it?